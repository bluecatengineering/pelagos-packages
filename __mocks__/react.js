module.exports = {
	...jest.requireActual('react'),
	useState: jest.fn((v) => [typeof v === 'function' ? v() : v, jest.fn()]),
	useCallback: (fn) => fn,
	useContext: jest.fn(),
	useEffect: jest.fn(),
	useLayoutEffect: jest.fn(),
	useMemo: jest.fn((f) => f()),
	useReducer: jest.fn((reducer, initial) => [initial, jest.fn()]),
	useRef: jest.fn((current) => ({current})),
	Children: {
		toArray(children) {
			return Array.isArray(children) ? children : Array.of(children);
		},
		map(children, fn) {
			return this.toArray(children).map(fn);
		},
	},
};
