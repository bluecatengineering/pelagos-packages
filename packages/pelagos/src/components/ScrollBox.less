@import '../../less/spacing';

.ScrollBox {
	flex-direction: row;
	flex: 1;
	overflow: hidden;
	align-items: center;
}

.ScrollBox__track {
	flex-direction: row;
	flex: 1;
	overflow: hidden;
	scroll-behavior: smooth;
}

.ScrollBox__btn {
	display: none;

	.ScrollBox--overflow > & {
		display: inherit;
	}

	@media print {
		.ScrollBox--overflow > & {
			display: none;
		}
	}
}
