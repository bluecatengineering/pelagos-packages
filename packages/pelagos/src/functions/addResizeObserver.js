/**
 * Adds a resize observer to the specified target element.
 * @param {Element} target the target element.
 * @param {function(DOMRectReadOnly): void} callback the function called when the element is resized.
 * @returns {function(): void} a function to disconnect the observer.
 *
 * @example
 * import {useCallback, useEffect, useRef} from 'react';
 * import {addResizeObserver} from '@bluecateng/pelagos';
 *
 * const Example = () => {
 *   const ref = useRef(null);
 *   const handleResize = useCallback(r => {
 *     // handle resize
 *   }, []);
 *   useEffect(() => addResizeObserver(ref.current, handleResize), [handleResize]);
 *   return <div ref={ref}>...</div>;
 * }
 */
const addResizeObserver = (target, callback) => {
	const observer = new ResizeObserver((entries) => callback(entries[0].contentRect));
	observer.observe(target);
	return () => observer.disconnect();
};

export default addResizeObserver;
