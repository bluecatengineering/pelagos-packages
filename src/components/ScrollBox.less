@import '../../less/spacing';
@import '../../less/utils';

.ScrollBox {
	flex-direction: row;
	flex: 1;
	overflow: hidden;
}

.ScrollBox__track {
	flex-direction: row;
	flex: 1;
	overflow: hidden;
	scroll-behavior: smooth;
}

.ScrollBox__btn {
	@icon-button();
	display: none;
	font-size: 14px;
	justify-content: center;
	padding: 0 @sp-04;
	margin: @sp-02;

	&:first-child {
		margin-right: @sp-08;
	}

	&:last-child {
		margin-left: @sp-08;
	}

	.ScrollBox--overflow > & {
		display: inherit;
	}

	@media print {
		.ScrollBox--overflow > & {
			display: none;
		}
	}
}
