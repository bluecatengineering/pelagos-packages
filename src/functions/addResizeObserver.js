import throttleAF from './throttleAF';

/**
 * Adds a resize observer to the specified target element.
 * @param {Element} target the target element.
 * @param {function(DOMRectReadOnly): void} callback the function called when the element is resized.
 * @returns {function(): void} a function to disconnect the observer.
 */
export default (target, callback) => {
	let rect;
	const f = throttleAF(() => callback(rect));
	const observer = new ResizeObserver((entries) => ((rect = entries[0].contentRect), f()));
	observer.observe(target);
	return () => (f.cancel(), observer.disconnect());
};
