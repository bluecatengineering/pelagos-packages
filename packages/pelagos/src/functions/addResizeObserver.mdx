{/* DO NOT EDIT, generated from addResizeObserver.js */}

import {Meta} from '@storybook/addon-docs/blocks';

<Meta title="functions/addResizeObserver" />

# addResizeObserver

Adds a resize observer to the specified target element.

## Parameters

<table>
	<thead>
		<tr>
			<th>Param</th>
			<th>Type</th>
			<th>Description</th>
		</tr>
	</thead>
	<tbody>
		<tr>
			<td>target</td>
			<td>`Element`</td>
			<td>the target element.</td>
		</tr>
		<tr>
			<td>callback</td>
			<td>`function`</td>
			<td>the function called when the element is resized.</td>
		</tr>
	</tbody>
</table>

## Returns

`function` - a function to disconnect the observer.

## Example

```js
import {useCallback, useEffect, useRef} from 'react';
import {addResizeObserver} from '@bluecateng/pelagos';

const Example = () => {
	const ref = useRef(null);
	const handleResize = useCallback((r) => {
		// handle resize
	}, []);
	useEffect(() => addResizeObserver(ref.current, handleResize), [handleResize]);
	return <div ref={ref}>...</div>;
};
```
