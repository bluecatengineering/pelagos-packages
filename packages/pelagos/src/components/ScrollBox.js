import {useCallback, useEffect, useRef} from 'react';
import PropTypes from 'prop-types';
import ChevronLeft from '@carbon/icons-react/es/ChevronLeft';
import ChevronRight from '@carbon/icons-react/es/ChevronRight';

import addResizeObserver from '../functions/addResizeObserver';

import IconButton from './IconButton';
import './ScrollBox.less';

const preventDefault = (event) => event.preventDefault();

const showChevrons = (track, onResize) => {
	const rect = track.getBoundingClientRect(); // the observer rectangle is incomplete
	const classList = track.parentNode.classList;
	const overflow = rect.width < track.firstChild.scrollWidth;
	if (overflow) {
		classList.add('ScrollBox--overflow');
	} else {
		classList.remove('ScrollBox--overflow');
	}
	if (onResize) onResize(track, rect, overflow);
};

/** Allows to scroll a nested component. */
const ScrollBox = ({className, trackId, children, onResize}) => {
	const trackRef = useRef(null);

	const handleClickLeft = useCallback(() => (trackRef.current.scrollLeft += -160), []);
	const handleClickRight = useCallback(() => (trackRef.current.scrollLeft += 160), []);

	useEffect(() => addResizeObserver(trackRef.current, () => showChevrons(trackRef.current, onResize)), [onResize]);
	useEffect(() => {
		showChevrons(trackRef.current, onResize);
	});

	return (
		<div className={`ScrollBox${className ? ` ${className}` : ''}`}>
			<IconButton
				className="ScrollBox__btn"
				icon={ChevronLeft}
				tabIndex="-1"
				aria-hidden="true"
				onMouseDown={preventDefault}
				onClick={handleClickLeft}
			/>
			<div id={trackId} className="ScrollBox__track" ref={trackRef}>
				{children}
			</div>
			<IconButton
				className="ScrollBox__btn"
				icon={ChevronRight}
				tabIndex="-1"
				aria-hidden="true"
				onMouseDown={preventDefault}
				onClick={handleClickRight}
			/>
		</div>
	);
};

ScrollBox.propTypes = {
	/** The component class name(s). */
	className: PropTypes.string,
	/** The ID for the track element. */
	trackId: PropTypes.string,
	/** The nested element. */
	children: PropTypes.element,
	/** Function invoked when the element is resized. */
	onResize: PropTypes.func,
};

export default ScrollBox;
