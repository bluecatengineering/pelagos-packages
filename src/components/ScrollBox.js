import {useCallback, useEffect, useRef} from 'react';
import PropTypes from 'prop-types';
import {t} from '@bluecat/l10n.macro';
import {faChevronLeft, faChevronRight} from '@fortawesome/free-solid-svg-icons';

import handleButtonKeyDown from '../functions/handleButtonKeyDown';
import addResizeObserver from '../functions/addResizeObserver';

import SvgIcon from './SvgIcon';
import './ScrollBox.less';

const preventDefault = (event) => event.preventDefault();

const showChevrons = (track, onResize) => {
	const rect = track.getBoundingClientRect(); // the observer rectangle is incomplete
	const classList = track.parentNode.classList;
	const overflow = rect.width < track.firstChild.scrollWidth;
	if (overflow) {
		classList.add('ScrollBox--overflow');
	} else {
		classList.remove('ScrollBox--overflow');
	}
	if (onResize) onResize(track, rect, overflow);
};

/** Allows to scroll a nested component. */
const ScrollBox = ({className, trackId, children, onResize}) => {
	const trackRef = useRef(null);

	const handleClickLeft = useCallback(() => (trackRef.current.scrollLeft += -160), []);
	const handleClickRight = useCallback(() => (trackRef.current.scrollLeft += 160), []);

	useEffect(() => addResizeObserver(trackRef.current, () => showChevrons(trackRef.current, onResize)), [onResize]);
	useEffect(() => {
		const track = trackRef.current;
		showChevrons(track, onResize);
	});

	return (
		<div className={`ScrollBox${className ? ` ${className}` : ''}`}>
			<div
				className="ScrollBox__btn"
				tabIndex="0"
				role="button"
				aria-label={t`Scroll left`}
				onMouseDown={preventDefault}
				onClick={handleClickLeft}
				onKeyDown={handleButtonKeyDown}>
				<SvgIcon icon={faChevronLeft} />
			</div>
			<div id={trackId} className="ScrollBox__track" ref={trackRef}>
				{children}
			</div>
			<div
				className="ScrollBox__btn"
				tabIndex="0"
				role="button"
				aria-label={t`Scroll right`}
				onMouseDown={preventDefault}
				onClick={handleClickRight}
				onKeyDown={handleButtonKeyDown}>
				<SvgIcon icon={faChevronRight} />
			</div>
		</div>
	);
};

ScrollBox.propTypes = {
	/** The component class name(s). */
	className: PropTypes.string,
	/** The ID for the track element. */
	trackId: PropTypes.string,
	/** The nested element. */
	children: PropTypes.element,
	/** Function invoked when the element is resized. */
	onResize: PropTypes.func,
};

export default ScrollBox;
