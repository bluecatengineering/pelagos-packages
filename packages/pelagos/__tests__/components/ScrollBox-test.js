import {useEffect, useRef} from 'react';
import {shallow} from 'enzyme';

import ScrollBox from '../../src/components/ScrollBox';
import addResizeObserver from '../../src/functions/addResizeObserver';

jest.unmock('../../src/components/ScrollBox');

describe('ScrollBox', () => {
	describe('rendering', () => {
		it('renders expected elements', () => {
			const wrapper = shallow(
				<ScrollBox>
					<div>Test</div>
				</ScrollBox>
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders expected elements when className is set', () => {
			const wrapper = shallow(
				<ScrollBox className="TestClass">
					<div>Test</div>
				</ScrollBox>
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});
	});

	describe('behaviour', () => {
		it('scrolls left when the left arrow is clicked', () => {
			const current = {scrollLeft: 500};
			useRef.mockReturnValue({current});
			const wrapper = shallow(<ScrollBox />);
			wrapper.find('IconButton').first().simulate('click');
			expect(current.scrollLeft).toBe(340);
		});

		it('scrolls right when the right arrow is clicked', () => {
			const current = {scrollLeft: 500};
			useRef.mockReturnValue({current});
			const wrapper = shallow(<ScrollBox />);
			wrapper.find('IconButton').last().simulate('click');
			expect(current.scrollLeft).toBe(660);
		});

		it('calls preventDefault on mouse down on the left arrow', () => {
			const preventDefault = jest.fn();
			const wrapper = shallow(<ScrollBox />);
			wrapper.find('IconButton').first().simulate('mousedown', {preventDefault});
			expect(preventDefault.mock.calls).toEqual([[]]);
		});

		it('calls preventDefault on mouse down on the right arrow', () => {
			const preventDefault = jest.fn();
			const wrapper = shallow(<ScrollBox />);
			wrapper.find('IconButton').last().simulate('mousedown', {preventDefault});
			expect(preventDefault.mock.calls).toEqual([[]]);
		});

		it('adds and removes resize observer', () => {
			const onResize = jest.fn();
			const rect = {width: 50};
			const current = {
				getBoundingClientRect: jest.fn().mockReturnValue(rect),
				firstChild: {scrollWidth: 100},
				parentNode: {classList: {add: jest.fn()}},
			};
			useRef.mockReturnValue({current});
			shallow(<ScrollBox onResize={onResize} />);
			expect(useEffect).toHaveBeenCalledTimes(2);
			useEffect.mock.calls[0][0]();
			expect(addResizeObserver.mock.calls).toEqual([[current, expect.any(Function)]]);
			addResizeObserver.mock.calls[0][1]();
			expect(current.parentNode.classList.add.mock.calls).toEqual([['ScrollBox--overflow']]);
			expect(onResize.mock.calls).toEqual([[current, rect, true]]);
		});

		it('removes chevrons when width is less', () => {
			const rect = {width: 100};
			const current = {
				getBoundingClientRect: jest.fn().mockReturnValue(rect),
				firstChild: {scrollWidth: 50},
				parentNode: {classList: {remove: jest.fn()}},
			};
			useRef.mockReturnValue({current});
			shallow(<ScrollBox />);
			useEffect.mock.calls[1][0]();
			expect(current.parentNode.classList.remove.mock.calls).toEqual([['ScrollBox--overflow']]);
		});
	});
});
