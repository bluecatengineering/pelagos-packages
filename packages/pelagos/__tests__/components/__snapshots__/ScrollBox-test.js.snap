// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollBox rendering renders expected elements 1`] = `
<div
  className="ScrollBox"
>
  <IconButton
    aria-hidden="true"
    className="ScrollBox__btn"
    icon={
      {
        "icon": [],
        "iconName": "chevron-left",
        "prefix": "fas",
      }
    }
    onClick={[Function]}
    onMouseDown={[Function]}
    size="medium"
    tabIndex="-1"
    tooltipPlacement="right"
    type="ghost"
  />
  <div
    className="ScrollBox__track"
  >
    <div>
      Test
    </div>
  </div>
  <IconButton
    aria-hidden="true"
    className="ScrollBox__btn"
    icon={
      {
        "icon": [],
        "iconName": "chevron-right",
        "prefix": "fas",
      }
    }
    onClick={[Function]}
    onMouseDown={[Function]}
    size="medium"
    tabIndex="-1"
    tooltipPlacement="right"
    type="ghost"
  />
</div>
`;

exports[`ScrollBox rendering renders expected elements when className is set 1`] = `
<div
  className="ScrollBox TestClass"
>
  <IconButton
    aria-hidden="true"
    className="ScrollBox__btn"
    icon={
      {
        "icon": [],
        "iconName": "chevron-left",
        "prefix": "fas",
      }
    }
    onClick={[Function]}
    onMouseDown={[Function]}
    size="medium"
    tabIndex="-1"
    tooltipPlacement="right"
    type="ghost"
  />
  <div
    className="ScrollBox__track"
  >
    <div>
      Test
    </div>
  </div>
  <IconButton
    aria-hidden="true"
    className="ScrollBox__btn"
    icon={
      {
        "icon": [],
        "iconName": "chevron-right",
        "prefix": "fas",
      }
    }
    onClick={[Function]}
    onMouseDown={[Function]}
    size="medium"
    tabIndex="-1"
    tooltipPlacement="right"
    type="ghost"
  />
</div>
`;
